packages/mongodb-log-writer/src/mongo-log-manager.ts

```diff
@@ -67,32 +84,57 @@ export class MongoLogManager {
       const { id } =
         /^(?<id>[a-f0-9]{24})_log(\.gz)?$/i.exec(dirent.name)?.groups ?? {};
       if (!id) continue;
+
       const fileTimestamp = +new ObjectId(id).getTimestamp();
       const fullPath = path.join(dir, dirent.name);
-      let toDelete: string | undefined;
 
       // If the file is older than expected, delete it. If the file is recent,
       // add it to the list of seen files, and if that list is too large, remove
       // the least recent file we've seen so far.
       if (fileTimestamp < deletionCutoffTimestamp) {
-        toDelete = fullPath;
-      } else if (this._options.maxLogFileCount) {
-        leastRecentFileHeap.push({ fullPath, fileTimestamp });
-        if (leastRecentFileHeap.size() > this._options.maxLogFileCount) {
-          toDelete = leastRecentFileHeap.pop()?.fullPath;
-        }
+        await this.deleteFile(fullPath);
+        continue;
       }
 
-      if (!toDelete) continue;
-      try {
-        await fs.unlink(toDelete);
-        // eslint-disable-next-line @typescript-eslint/no-explicit-any
-      } catch (err: any) {
-        if (err?.code !== 'ENOENT') {
-          // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
-          this._options.onerror(err, fullPath);
+      let fileSize: number | undefined;
+      if (this._options.retentionGB) {
+        try {
+          fileSize = (await fs.stat(fullPath)).size;
+          usedStorageSize += fileSize;
+        } catch (err) {
+          this._options.onerror(err as Error, fullPath);
+          continue;
         }
       }
+
+      if (this._options.maxLogFileCount || this._options.retentionGB) {
+        leastRecentFileHeap.push({ fullPath, fileTimestamp, fileSize });
+      }
+
+      if (
+        this._options.maxLogFileCount &&
+        leastRecentFileHeap.size() > this._options.maxLogFileCount
+      ) {
+        const toDelete = leastRecentFileHeap.pop();
+        if (!toDelete) continue;
+        await this.deleteFile(toDelete.fullPath);
+        usedStorageSize -= toDelete.fileSize ?? 0;
+      }
+    }
+
+    if (this._options.retentionGB) {
```

Comment:
I think this should address most of the concerns.
Not sure what other issues might be with `fs.stat` worth handling